### ISSUE TITLE
Requesting a net35 target for the S.R.Metadata package

### ISSUE THREAD

**Issue by jnm2** [dotnet/corefx#27469]
Projects like https://github.com/nunit/nunit-console (including the engine shared by the console and adapter) and https://github.com/OpenCover/opencover allow users to execute their assemblies on desktop CLR v2 and therefore cannot target net40 or greater.

Without a net35 target, or maintaining such a port, we can't move away from Mono.Cecil. Mono.Cecil a major drawback, crashing when reading portable PDB information. It hasn't been updated in a while. S.R.Metadata is also better-designed for efficient low-level reading.

This would make it desirable to publish a net35 build of S.C.Immutable. @sharwell has one at https://github.com/tunnelvisionlabs/dotnet-collections, though adding a net35 target to the S.C.Immutable package would also be welcome for even more projects than stated above.

It looks like this would make the difference for me between days and months of work, so I'm happy to contribute here!

**Comment by akoeplinger** [comment]
> Mono.Cecil has a major drawback for projects that use rely on it, crashing when reading portable PDB information.

Which version are you using? Portable PDB reading has been supported in Mono.Cecil since quite some time.

**Comment by jnm2** [comment]
The current version, 0.9.6.4. Versions of 0.10 have been in beta so far.

S.R.Metadata is the API I'd prefer to use since I use it elsewhere. It's low-level rather than eagerly building an object model.

**Comment by akoeplinger** [comment]
0.10 is the most "current" version right now on the road to 1.0, but yes it does have some API breakages :)

**Comment by jnm2** [comment]
If 0.10 wasn't published as prerelease, and it had a similar low-level API, that would be an interesting option.

**Comment by akoeplinger** [comment]
@jbevain do you recommend people use Cecil 0.10 or is 0.9 still the recommended version?

If you search for "Mono.Cecil 0.10 beta 1" on http://cecil.pe/ there's a list of the breaking changes, it doesn't introduce a low-level API like S.R.Metadata has though.

**Comment by tmat** [comment]
@jnm2 Do NUnit and OpenCover customers who run their tests on .NET 3.5 or 4.0 actively update to the latest version of these tools? What does such upgrade get them? Wouldn't it be possible to stop supporting these old platforms in newer versions of NUnit and OpenCover and port critical fixes to the older versions that still target .NET 3.5/4.0?

Porting and maintaining System.Collection.Immutable and System.Reflection.Metadata for .NET 3.5 would add significant complexity to the build (many `#ifdefs`) and a maintenance cost. We would need to justify this investment by data that show that a lot of customers would benefit from it.

**Comment by jnm2** [comment]
@tmat I can't speak for OpenCover. For NUnit, people have been upgrading for bugfixes and for new features. 

For instance, the company I work for runs net35 tests because we write modules for a financial platform which requires CLR v2. We want our tests to exercise our module code against the same runtime. That need won't go away any time soon, and yet, we wouldn't necessarily be happy being stuck at the current versions of all our dependencies. 
The NUnit VSTest adapter has a filtering bug which is causing us grief. I'm working on a partial fix for 3.10 with a larger fix to follow in 3.11. We will absolutely gobble that up in our CI scripts at work and remove the complications we've placed in our source code—so long as NUnit's engine still supports net35.

I know of others who have business requirements to develop on net40 and want us to keep that target.

Justifying the complexity is understandable. I would help collect such data if I could? Looking at NuGet, I see unofficial backports for net40 for [S.C.Immutable](https://www.nuget.org/packages/System.Collections.Immutable.Net40/) and [S.R.Metadata](https://www.nuget.org/packages/Net40.Reflection.Metadata/) (ew, no project site?), besides Sam's port of [S.C.Immutable](https://www.nuget.org/packages/Rackspace.Collections.Immutable) for net35.

**Comment by jnm2** [comment]
If I remember correctly, @masonwheeler who maintains the Boo language was looking for a pre-net46 System.Reflection.Metadata.

**Comment by jnm2** [comment]
If I remember correctly, @masonwheeler who maintains the Boo language was looking for a pre-net46 System.Reflection.Metadata.

If I also heard correctly, the ILSpy team wants to move to S.R.Metadata but they are also targeting pre-net46.

**Comment by tmat** [comment]
@jnm2 Do you run your unit test on a VM that has net35?

**Comment by jnm2** [comment]
@tmat Yes. VSTest has an assembly specifically compiled against net35 so that it can host tests on CLR v2. NUnit's engine does the same.

**Comment by jnm2** [comment]
@tmat Yes. VSTest has an executable specifically compiled against net35 so that it can host tests on CLR v2. NUnit's engine does the same.

CLR v2 and v4 install side-by-side, unlike the in-place updates that have happened since net40. But with net45 and up, even there quirking comes into play. We just got burned by net45 vs net46 quirking in the real world this month.

**Comment by masonwheeler** [comment]
@jnm2 
>If I remember correctly, @masonwheeler who maintains the Boo language was looking for a pre-net46 System.Reflection.Metadata.

That's right.  The compiler currently runs against a minimum of v4.5, and I'd prefer not to have to bump the minimum dependency if possible.
